A few changes in error reporting on non-existent files in coqdep, ocamllibdep, coq_makefile, including #11631 and #14539

herbelin opened the pull request:
**Kind:** fix / enhancements

This PR addresses the following:
- spurious clash warnings with `From foo Require bar`: this is a hot fix while waiting for @ejgallgo's loadpath cleanup  (#11631, #14539)
- anomalies or internal errors with non-existent directories and paths:
  - `ocamllib`: changing `Uncaught exception: Sys_error("foo: No such file or directory")` into `*** Warning: foo: No such file or directory` on non-existent include (should it be an error or a warning?)
  - `coqdep` on non-existent file, after #14024 which fixed a silent ignoring: from `File "foo", line 1: No such file or directory` to `Error: foo: No such file or directory` from the command line, or `*** Warning: In "_FooProject": bar: No such file or directory` if indirectly through a `_CoqProject`
   - `coqdep` on non-existent directory: from a "please report" anomaly to `*** Warning: cannot open foo` (should it be an error?)
  - `coqdep`: when a file exists but not with the expected `From root`, clarify the message that it does not exist only wrt this root
  - `coq_makefile` on non-existent project file: changing `Fatal error: exception Sys_error("foo: No such file or directory")` into `Error: foo: No such file or directory`

Among others, fixes #11631, fixes #14539.

Not necessarily clear when it should error or warn. Not clear also which exact format to adopt (currently, e.g., it can be with a `***` prefix or not).

- [x] Added / updated **test-suite**.

- [ ] Added **changelog**.

ppedrot commented:
> this is a hot fix

@herbelin last time I had a look the code was completely nonsensical. I don't think your code restores soundness at all, it's just hiding the nonsense under the rug and is probably going to break some developments in fancy ways. I'd recommend kicking @ejgallego 's ass for the PR development to go faster rather than piling hacks over hacks in this very broken part of the code. Worst case that qualifies as a reasonable hotfix is to copy Loadpath into coqdep and remove it when the factoring actually happens.

herbelin commented:
Interestingly, the PR reveals that in the presence of overlapping between `-R`/`-Q` options, there are different kinds of overlapping. Let's consider the following layout:
```
├── B1
│   └── C
│   |   └── E.v
│   └── D
│       └── E.v
├── B2
    └── E.v
```
In `coqdep -R B1 B -R B2 B`, there are ambiguities of different kinds: the resolution of `E` between `B.C.E` and `B.D.E` is determined by the order between `C` and `D` in the file system (knowing that the underlying hash table fortunately preserves the order of entry in its buckets), but the discrimination with `B.E` depends only in the order of arguments on the command line. So, we may like to warn only in the first case and not in the second case.

This is however not what `coqdep` does: it always warns and does not take into account that the order in the command line gives a criterion to decide even though both it and `coqc` actually use this criterion to take a decision.

herbelin commented:
> the code was completely nonsensical. I don't think your code restores soundness at all...

What kind of "nonsensical" do you have in mind? Barring corner cases, it seems these days to reasonably work? (I'm not saying that the code should not be shared with `coqc`, on the contrary, but it is relatively better now than say 10 years ago, at a time where `coqc` and `coqdep` quickly disagreed).

Moreover, the warnings I added (as a compensation of the ones I removed) allow to reveal that e.g. hierarchy-builder was talking some risks in having two files named the same in the same file tree, w/o probably having realized they were doing it.

SkySkimmer commented on lib/coqProject_file.ml:
```diff
@@ -116,12 +116,11 @@ and parse_args buf accu s = match Stream.next s with
 
 let parse f =
-  let c = open_in f in
+  let c = try open_in f with Sys_error msg -> Format.eprintf "Error: @[%s@].@\n%!" msg; exit 1 in
```
Do we really want this to be unrecoverable? Seems like it should just raise a Parsing_error

herbelin replied:
You mean returning a parsing error if the project file to open cannot be opened? This looks confusing to me.

SkySkimmer replied:
Better than exit imo

Alizter replied:
Perhaps we need an `UnableToOpenProjectFile` exception that reraises this exception. That way it is clear that the project file was unable to be opened and the reason why. I agree that exiting is probably not the best thing to do.

SkySkimmer replied:
I don't see what's confusing about `Parsing_error "Could not open project file foo"` tbh
alternatively we don't need to catch the sys_error

herbelin replied:
As `coqdep` is currently done, it reports uncaught errors as anomalies, so we need to do something. I'd be tempted to adopt the `UnableToOpenProjectFile` proposal.

Alizter commented on tools/coqdep_common.ml:
```diff
@@ -127,60 +129,74 @@ let get_prefix p l =
   in
-  try Hashtbl.iter iter table; raise Not_found
-  with M.Found s -> s
+  (* We reverse to give priority for compatibility. There is an issue
```
Can you expand on the compatibility requirements. Is anyone relying on this behaviour?

herbelin replied:
hierarchy-builder was relying on this order.

SkySkimmer reviewed (changes requested):
Do not `exit 1`

herbelin commented:
> Do not `exit 1`

I added an exception `UnableToOpenProjectFile` but we still need to eventually do something when the exception is raised. I don't know what else to do than reporting a failure and exiting. Tell me if you have something else in mind.

ppedrot commented:
It looks like this PR is turning into a hodgepodge of random fixes that are loosely related and not even for the same binaries (`coq_makefile` vs. `coqdep`). Couldn't we get instead smaller PRs that would be merged faster and more easily?

Also, my main remark still stands: coqdep is just as broken as before and no amount of haphazard fiddling with the set of clashes is going to fix it. You need to completely rework the thing for it to be sound.

herbelin commented:
Yes, it is a hodgepodge of not random but indeed heteroclitous fixes while observing various problems of coqdep, ocamllibdep and coq_makefile.

Sure, it can be split. Will do.

About reworking the whole thing, it will not be by me, but I will not cry for the current coqdep if someone finds the courage to rewrite it.

herbelin commented:
As another divergence between `coqdep` and `coqc`, in:
```
├── B
│   └── D.v  (* 1 *)
│   └── C
│       └── D.v (* 2 *)
```
`coqc` treats the files of `B` before the ones of `B/C` leading to give priority to instance 1 of `D.v` while `coqdep` treats files and subdirectories in a same pass, so depending on the internal order of the contents of B, treating 1 first if `D.v` comes before `C` in the order given by `Sys.readdir` for directory `B` and treating 2 first otherwise.

herbelin commented:
> I'd recommend kicking @ejgallego 's ass for the PR development to go faster rather than piling hacks over hacks in this very broken part of the code. Worst case that qualifies as a reasonable hotfix is to copy Loadpath into coqdep and remove it when the factoring actually happens.

@ejgallego, @ppedrot: how would you rewrite `coqdep`? In addition to preserving order, I believe efficiency is important, no? Would `dune` comes with support for what we need? Otherwise, would you typically have in mind a tree-like structure with access first from the top following the qualifiers of the `From` field then going from the bottom following the qualifiers of the module to find, as e.g.:
```
type entry = file_name * logical_name * original_include_command
type ('a,'b) tree = Leaf of 'a | Node of 'b
type rev_dict = (entry list, rev_dict) tree StringMap.t
type dict = (rev_dict, dict) tree StringMap.t
```
In any case, this would not prevent being cautious about the order files are added to the structure, deciding when to override and when to report a clash.

I'm afrait including `Loadpath` would be costly...

herbelin commented:
Another parameter to take into account about ordering:
```
├── B
│   └── E.v (* 1 *)
│   └── C
│       └── E.v (* 2 *)
│       └── D
│           └── E.v (* 3 *)
```
do we go upwards or downwards? `coqc` goes in a mixed order 1, 3, 2 and, probably, we should keep this order for compatibility. Or we can also adopt the more intuitive order 1, 2, 3? The reason for the mixed order is that the top directory is treated separately. So with more nestings, it would be 1, n+1, n, ..., 2.

ejgallego commented:
@herbelin indeed we likely need to tweak the current semantics a bit, but we need to put all the loadpath code in one place first.

ejgallego commented:
@herbelin thanks for all the careful and detailed analysis here, it is indeed super useful.

herbelin commented:
> @herbelin indeed we likely need to tweak the current semantics a bit, but we need to put all the loadpath code in one place first.

One thing about sharing the loadpath code is that `coqc` works in a lazy way in the sense that it builds first the loadpath structure without looking at the files, then, for each `Require`, looks (dynamically) in the different possible directories. I don't know how much this is really useful, but, for instance, you can compile files after having started the editor and have the files taken into account only dynamically.

On the other side, mostly for efficiency reasons, `coqdep` does not build a loadpath structure but instead directly build a map binding qualified names (or rather pairs of from qualifier and require qualifiers) to filenames.

To have all the loadpath code in one place, do you lean towards relying on the coqc model or on the coqdep model?

herbelin commented:
I gave in to and implemented a direct map from pairs `(from,suffix)` with a hash-table. Execution times seem equivalent (that is, most of the time is spent in building the `-R coqlibpath Coq` map!).

This allows to be independent of the order of `Hashtbl.fold` and to fix the [second part](https://github.com/coq/coq/issues/14539#issuecomment-895501948) of #14539. This also somehow answers a comment of @ppedrot in 31c754273 ("We should rather use some trie-based data structure, but I'll leave this for a future commit."). [Note that this does not prevent at all a cleanup of loadpath management in general, on the contrary. The knowledge we acquired in the process will be useful for a cleanup.]

I added more tests showing how ambiguities are resolved when a file occurs several times in the same `-Q`/`-R` subspace showing the dependency on the file system (that is on the order given by `Sys.readdir`). As a result, the Windows test and my own setup show a different resolution of the ambiguity than the MacOS and base tests! I then don't see what alternatives there are than forbidding such `Require`s whose resolution rely on the filesystem order??

ppedrot commented:
>  I then don't see what alternatives there are than forbidding such Requires whose resolution rely on the filesystem order??

Why not normalize the output of `readdir`? The OCaml doc explicitly states that
> There is no guarantee that the name strings in the resulting array will appear in any specific order; they are not, in particular, guaranteed to appear in alphabetical order.

herbelin commented:
> Why not normalize the output of readdir?

You mean using byte order? Yes, that's a possibility also.

There might be utf8 issues though when working across different platforms. A same unicode file can be encoded in different utf8 ways depending on file systems (e.g. on MacOS X, accented letters are encoded in filenames with combining accents).

ppedrot commented:
> You mean using byte order?

Not necessarily with byte order. As you mention, we could also perform Unicode normalization. That said, it's not clear that we wouldn't already have other similar issues with cross-platform binary vo files in coqc, since we handle identifiers as mere binary strings.

herbelin commented:
> > You mean using byte order?
> 
> Not necessarily with byte order. As you mention, we could also perform Unicode normalization.

You mean Unicode normalization + `Stdlib.compare`? Or collation? The first is maybe easier according to [uucp](https://erratique.ch/software/uucp/doc/unicode.html#eqcmpnorm)'s page.

> That said, it's not clear that we wouldn't already have other similar issues with cross-platform binary vo files in coqc, since we handle identifiers as mere binary strings.

I believe we embed strings the form they were entered, even the few filenames stored in locations.

As for filenames, (now old) HFS+ for instance [unicode-normalizes](https://developer.apple.com/library/archive/technotes/tn/tn1150.html#UnicodeSubtleties) filenames before doing comparison. New APFS does not normalize itself but MacOS libraries apparently normalize before communicating with APFS, according to this [page](https://eclecticlight.co/2021/05/08/explainer-unicode-normalization-and-apfs/).

Incidentally, should we normalize the kernel names, or at least normalize the qualids before asking the nametab?

ejgallego commented:
Sorry folks I am not following the discussion closely, but something I am positive based on experience with large builds is that all directory traversal has to be lazy if possible.

herbelin commented:
> Sorry folks I am not following the discussion closely, but something I am positive based on experience with large builds is that all directory traversal has to be lazy if possible.

As a matter of fact, most of the time spent in a call to `coqdep` is to reconstruct the map for the standard library. It should indeed be done lazily or maybe even cached somewhere.

herbelin commented:
As far as I can judge, the error in `perennial` was "real": `coqdep` and `coqc` disagreed on how to manage the loadpath but since, thanks to `coqdep`, only one of the two conflicting files was compiled, the vo file that `coqc` would have taken by default was absent and the file mentioned by `coqdep` was taken.

Told in another way, `coqdep` may impact the choices made by `coqc` by imposing an order of compilation to `coqc`.

So, it is not clear that `coqdep` should necessarily comply with `coqc`. It can be `coqc` which align on `coqdep`.

The last commit I added makes that `coqc` follows `coqdep` on the order in which subdirectories are considered (a `List.rev` is added).

(The problem of `Sys.readdir`'s order being platform or file-system dependent remains though.)

herbelin commented on test-suite/misc/deps-order-from.sh:
```diff
@@ -0,0 +1,22 @@
+$coqc -R misc/deps/test-from T misc/deps/test-from/E.v 2>&1
+S=$?
+if [ $R = 0 ] && [ $S = 0 ]; then
+    printf "coqdep and coqc agree\n"
+    exit 0
+else
+    printf "coqdep and coqc disagree\n"
+    exit 1
```
You mean adding two periods right?

[BTW, note that the message does not exactly capture agreement/disagreement but also failure of one and not failure of the other or such kind of things. I did not take time to fully investigate what the message captured and I left the current approximation.]

SkySkimmer commented on test-suite/misc/deps-order-from.sh:
```diff
@@ -0,0 +1,22 @@
+$coqc -R misc/deps/test-from T misc/deps/test-from/A/C.v 2>&1
+$coqc -R misc/deps/test-from T misc/deps/test-from/B/C.v 2>&1
+$coqc -R misc/deps/test-from T misc/deps/test-from/D.v 2>&1
+$coqc -R misc/deps/test-from T misc/deps/test-from/E.v 2>&1
```
Why all the `2>&1`?

herbelin replied:
If there are warnings, we may want to know about them, no?

Alizter reviewed (commented):
Just this comment about exceptions for now. Will review more tomorrow.

herbelin commented on tools/coqdep_common.ml:
```diff
@@ -115,45 +109,22 @@ let mkknown () =
-  try Hashtbl.iter iter table; raise Not_found
-  with M.Found s -> s
+let search_table table ?(from=[]) s =
+  Hashtbl.find table (from, s)
```
I kept it as it was, but we could use `find_opt` as well. I can do it if you insist.

Alizter commented:
I will merge tomorrow if @ppedrot is happy with it

herbelin commented:
Note that for `From`, it uses now a direct access in a hash table while before it was looking in a list. So, I suspect that performance can only be better in general.

Note also that performance-wise, a lot of the time is spent to reconstruct again and again the file map for the stdlib, so a cache would help here if someone knows how to build one (for coqdep on a single file, rebuilding the stdlib map is close to 100% of the time).

herbelin commented:
> Bench indicates no issue

Aren't times around 0 on the average in general? (Anyway, I don't see a reason why coquelicot would be slower because of coqdep.)

herbelin commented:
I finally implemented a variant of what was discussed at [Zulip](https://coq.zulipchat.com/#narrow/stream/237656-Coq-devs.20.26.20plugin.20devs/topic/Directory.20handling.20code.20in.20coqdep):

- If some files exactly match the required logical name, they are selected. If there is a `From` clause, this uniquely determines a file. If there is no `From` clause, the file corresponding to the most recently given `-R` or `-Q` is selected.
- If no files exactly match the required logical name, the ones corresponding to the most recently given `-R` or `-Q` (or to the `From` prefix if any) are selected. If there are several of them (in case of `-R`), the choice is system-dependent. [It could be possible to raise an error with a little extra effort in `coqc`, but I did not make this effort yet.]

This change of semantics seems rather natural and it provides a way to non-ambiguously refer to a given file, thus answering to this [Coq Club discussion](https://sympa.inria.fr/sympa/arc/coq-club/2020-09/msg00043.html). It is also compatible with the CI.

Alizter commented:
@herbelin Can you have a look at Jim's review? I want to merge this soon, but we should finish polishing up the doc.

herbelin commented:
> > I'm now also leaning in the direction to fail if a `Require` is ambiguous.
> 
> A warning (and ignoring one file) would be more user-friendly.

You mean as soon as `coqdep`?

> The conflict may be unimportant. The message should be specific about the conflicting files.

Yes, and the code for reporting the conflict is already done (with a warning).

> If there are several conflicts, they should all be reported. I expect `coqc` exits after reporting the first error, which would be less friendly.

This is one of the reason which pushes me to prefer a warning in `coqdep` (but the warning should not be the current warning, it should be something about "ambiguous path, it would fail at the time of compiling" or something like that)

> > The order in which subdirectories of `user-contrib` are added to the loadpath is dependent on the file system.
> 
> How about sorting the list of filenames in each directory before loading them? That would make it not system dependent. Sorting unicode correctly (from the user perspective) is locale-dependent, so that's not a good idea. Sorting on Unicode in some fixed locale or just doing a byte comparision of the Unicode would be sufficient to make it not system dependent.

Sorting with Unicode was discussed [above](https://github.com/coq/coq/pull/14718#issuecomment-913134938). One would have to rely on a unicode library, and it'd be a bit of work to implement, but why not.

> > if there are `Foo1` and `Foo2` in `user-contrib`, and both contains a file `Bar`, `Require Bar` will be system dependent. One answer would be to force to use `From` whenever we require a file from an installed package? Is it acceptable?
> 
> The shorter form is nice for users. Perhaps the code could generate an error only when `Bar` is ambiguous?

I was suggesting a `From`, because it would then be robust to addition of new possibly-newly-conflicting libraries in `user-contrib`.

SkySkimmer commented:
>Option -Q does not require From, if this is what you mean.

I mean for instance I cannot do `Require Ltac2` I need to do `Require Ltac2.Ltac2` or `From Ltac2 Require Ltac2`, so even if someone installs `user-contrib/foo/Ltac2.vo` there is no confusion.

jfehrle commented:
OTOH, it occurs to me that if the Coq strongly discourages having ambiguous files, then maybe Unicode sorting isn't really necessary.

herbelin commented:
Another issue with Unicode is that the loadpath is built by looking at directory names as they are in the file system (that is Unicode-normalized if on a file system that normalizes), but the correspondence with the logical path is done without normalization. Either a normalization function should be used (with applicability to all file systems) or the normalization function of the underlying file system (if any) is used to normalize (e.g. by creating a tmp file and reading its normalized name back).

herbelin commented:
Sorry for yet an iteration on the PR. I modified the following behaviors:
- `coqc`/`coqtop` now fail if a logical path partially matches two files in the same `-R`/`-Q` root; on its side, `coqdep` warns that `Require` will fail (maybe `coqdep` should already fail as it currently produces dependency files which are file system and that's a bit strange)
- `coqc`/`coqtop` do not any longer warn when two exact matches exist; instead, they take the latest; `coqdep` warns that the latest will be used
- also refined `Load`: `Load id` searches in path, but `Load "basename/filename"`, which is a non-sense in terms of logical path, does not

Note that, eventually, a few commits will have to be smashed. I'm also open to split the PR again in several smaller PRs if it helps.

herbelin commented:
> Any clue what the issue with equations is?

It is an ambiguous `Require`, see mattam82/Coq-Equations#435.

herbelin commented:
I squashed, making the different parts of the PR more visible:
- Fix of #11631 (initial report) and #14539
- Misc internal cleaning/renaming in coqdep_common.ml
- Adopting same order for browsing directories in coqc and coqdep
- A mini-fix in an error message introduced in #14827
- Removing an outdated comment in the refman
- An update of coqdep's man page and refman
- The new exact-match priority, order on `-R`/`-Q`, failure on ambiguity rules

ejgallego commented:
This seems good to go, great effort by @herbelin !

We have many options on how to merge the two codepaths, there is not a canonical strategy IMHO, as of today, I think that just make `coqdep` directly depend on `Loadpath`, even if `Loadpath`still depends on `coq-core.library` is the best way to go.

For that you need a small commit, see: https://github.com/ejgallego/coq/commit/495795f13e2c93f26ddfca8b391b585289a44242

herbelin commented:
> Caching is IMHO an orthogonal issue, tries could support both [and I was under the impression we already do caching in the `coqc` case, don't we @herbelin] ?

There is (or at least there was at some time) a caching of loaded `vo` files, but looking for an existing `vo` files was only at the time of requiring.

I don't have a strong opinion on the data structures to use, otherwise. Feel free to experiment what you think is good.
